### DISCUSSION TITLE
How do you compare to typed-inject and other alternatives ?

### DISCUSSION THREAD

**Discussion by jpbourgeon** [vovaspace/brandi#18] [Q&A] [answered]
Hi,

I'm in the process of selecting a Typescript DI Container and I would like to know how brandi compares to typed-inject, which was my choice before I crossed brandi's road ? The characteristics you put forward seem quite similar :

- Fully typed
- No decorators
- very lightweight

The noticeable differences to me are :
- You provide a React extension
- typed-inject is more mature, its community is more developed

What missing feature in other DI containers motivated you to develop brandi ?

My question is sincere. I do not intend to question the legitimity of your library. I just want to find the best DI container for my need.

Best regards

**Comment by vovaspace** [comment, marked as answer]
Hi. Thank you for your question.

I took the time to get a better understanding of Typed Inject, and here are the differences I saw:

- The Brandi tokens mechanism. Typed Inject suggests resolving dependencies in this way: `const myService = appInjector.injectClass(MyService);`. In this case, your code still depends on the `MyService` class constructor, instead of depending on the typed token as Brandi suggests: `const myService = container.get(myServiceToken)`;
- Brandi provides a different mechanism for [hierarchical containers](https://brandi.js.org/reference/hierarchical-containers). It fits perfectly with the hierarchical structure of the React-components;
- Both libraries have two identical dependency scopes: Transient and Singleton. However, Brandi has two additional scopes: Container and Resolution (It is especially useful to use Сontainer scope [together with React components](https://brandi.js.org/brandi-react/container-provider));
- I didn't see if it was possible to use [optional dependencies](https://brandi.js.org/reference/optional-dependencies) in Typed Inject. It may be useful, for example, to disable a logger in the production environment;
- Brandi provides a possibility to organize bindings into [dependency modules](https://brandi.js.org/reference/dependency-modules). This feature is stable, but I'm still not sure about the design. It will be great if you have ideas how to improve it.


> typed-inject is more mature, its community is more developed


You're right here, but since Brandi has fewer users, I'm freer to add new features and users have more power to influence the development of the library (yours PRs welcome).

I hope I answered your question.

**Reply by jpbourgeon** [reply]
Thank you for your thoughtful answer !
This is definitely enlightening and I now have a better understanding of the additional features provided by the tool.
I will definitely give it a try !
